# COVID-19: Exploratory Data Analysis & Visualization

by **Jennie Tram Le**

Inspired by


### Part I: Environment Setup

In [1]:
# Install wget to download the data from github
#!pip install wget

# Install map visualization package (folium, altair)
#!pip install folium
#!pip install altair

# Install interactive visualization package (plotly)
#!pip install plotly

In [2]:
# Import necessary package
import os
import wget
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# seaborn: statistical data visualization
import seaborn as sns
sns.set_style("whitegrid")

# Configuration for high resolution for rendered images on notebook
%config InlineBackend.figure_format='retina'

# Map visualization
# folium: choropleth visualization
import folium
# altair: declarative statistical visualization
import altair as alt

# Plotly for interactive visualization
from plotly import tools, subplots
import plotly.offline as py
py.init_notebook_mode(connected=True)   # require to use plotly offline in jupyter notebook
import plotly.graph_objs as go
import plotly.express as px
import plotly.figure_factory as ff
import plotly.io as pio
pio.templates.default = "plotly_white"

'''Display markdown formatted output like bold, italic bold, ...'''
from IPython.display import Markdown
def bold(string):
    display(Markdown(string))

## Part II: Download Data

In [3]:
# Remove on the downloaded csv files
! del *.csv

# Link urls to the updated data from John Hopkins University
urls = [
    # global confirmed cases
    'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_global.csv',
    # global death cases
    'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_deaths_global.csv']

# download the data
for url in urls:
    file = wget.download(url)

In [4]:
# Double check if the data are in the folder
! dir *.csv

 Volume in drive C is Windows-SSD
 Volume Serial Number is 6094-DC3C

 Directory of C:\Users\Jennie\Desktop\JENNIE\FORDHAM\Self-Study\Data Visualization\Python

09/12/2020  03:38 PM           255,898 time_series_covid19_confirmed_global.csv
09/12/2020  03:38 PM           180,466 time_series_covid19_deaths_global.csv
               2 File(s)        436,364 bytes
               0 Dir(s)  138,108,227,584 bytes free


In [5]:
# Read the Global Confirmed Cases Dataset
cases = pd.read_csv('time_series_covid19_confirmed_global.csv')
cases.head()

Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  9/2/20  9/3/20  9/4/20  9/5/20  \
0        0        0        0        0  ...   38243   38288   38304   38324   
1        0        0        0        0  ...    9728    9844    9967   10102   
2        0        0        0        0  ...   45158   45469   45773   46071   
3        0        0        0        0  ...    1199    1199    1215    1215   
4        0        0        0        0  ...    2777    2805    2876    2935   

   9/6/20  9/7/20  9/8/20  9/9/20  9/10/20  9/11/20  
0   38398   38494   38520   38544    38572    38606  
1   10255   10406   10553   10704    10860    11021  
2   46364   46653   46938   47216    47488    47752  
3    1215    1261    1261    1301     1301     1344  
4    2965    2981    3033    3092     3217     3279  

[5 rows x 238 columns]

In [6]:
# Read the Death Confirmed Cases Dataset
deaths = pd.read_csv('time_series_covid19_deaths_global.csv')
deaths.head()

Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  9/2/20  9/3/20  9/4/20  9/5/20  \
0        0        0        0        0  ...    1409    1409    1409    1409   
1        0        0        0        0  ...     296     301     306     312   
2        0        0        0        0  ...    1523    1529    1539    1549   
3        0        0        0        0  ...      53      53      53      53   
4        0        0        0        0  ...     112     113     115     117   

   9/6/20  9/7/20  9/8/20  9/9/20  9/10/20  9/11/20  
0    1412    1415    1418    1420     1420     1420  
1     316     319     321     322      324      327  
2    1556    1562    1571    1581     1591     1599  
3      53      53      53      53       53       53  
4     117     120     124     126      130      131  

[5 rows x 238 columns]

## Part III: Exploratory Data Analysis


#### 1. A bird-view at the data

In [7]:
# Checking the data size
# confirmed cases
print("****GLOBAL CONFIRMED CASES*****")
print("# rows:", cases.shape[0])
print("# columns:", cases.shape[1])
print("****GLOBAL DEATHS*****")
print("# rows:", deaths.shape[0])
print("# columns:", deaths.shape[1])

****GLOBAL CONFIRMED CASES*****
# rows: 266
# columns: 238
****GLOBAL DEATHS*****
# rows: 266
# columns: 238


In [8]:
# checking columns name
print("****All Columns Global Confirmed Cases****:\n", cases.columns)
print("\n")
print("****All Columns Gobal Deaths****:\n", deaths.columns)

****All Columns Global Confirmed Cases****:
 Index(['Province/State', 'Country/Region', 'Lat', 'Long', '1/22/20', '1/23/20',
       '1/24/20', '1/25/20', '1/26/20', '1/27/20',
       ...
       '9/2/20', '9/3/20', '9/4/20', '9/5/20', '9/6/20', '9/7/20', '9/8/20',
       '9/9/20', '9/10/20', '9/11/20'],
      dtype='object', length=238)


****All Columns Gobal Deaths****:
 Index(['Province/State', 'Country/Region', 'Lat', 'Long', '1/22/20', '1/23/20',
       '1/24/20', '1/25/20', '1/26/20', '1/27/20',
       ...
       '9/2/20', '9/3/20', '9/4/20', '9/5/20', '9/6/20', '9/7/20', '9/8/20',
       '9/9/20', '9/10/20', '9/11/20'],
      dtype='object', length=238)


In [9]:
# checking data types
print("Global Confirmed Cases Types:\n", cases.dtypes)
print("\nGlobal Deaths Types:\n", deaths.dtypes)

Global Confirmed Cases Types:
 Province/State     object
Country/Region     object
Lat               float64
Long              float64
1/22/20             int64
                   ...   
9/7/20              int64
9/8/20              int64
9/9/20              int64
9/10/20             int64
9/11/20             int64
Length: 238, dtype: object

Global Deaths Types:
 Province/State     object
Country/Region     object
Lat               float64
Long              float64
1/22/20             int64
                   ...   
9/7/20              int64
9/8/20              int64
9/9/20              int64
9/10/20             int64
9/11/20             int64
Length: 238, dtype: object


In [10]:
# Statistical checking for Global Confirmed Cases
cases.describe(include='all')

Province/State Country/Region         Lat        Long     1/22/20  \
count              81            266  266.000000  266.000000  266.000000   
unique             81            188         NaN         NaN         NaN   
top           Sichuan          China         NaN         NaN         NaN   
freq                1             33         NaN         NaN         NaN   
mean              NaN            NaN   21.076624   22.829441    2.086466   
std               NaN            NaN   24.904600   70.191330   27.279200   
min               NaN            NaN  -51.796300 -135.000000    0.000000   
25%               NaN            NaN    6.677575  -15.277675    0.000000   
50%               NaN            NaN   22.820850   20.921188    0.000000   
75%               NaN            NaN   41.143200   77.913685    0.000000   
max               NaN            NaN   71.706900  178.065000  444.000000   

           1/23/20     1/24/20     1/25/20      1/26/20      1/27/20  ...  \
count   266.000000  266.000000  266.000000   266.000000   266.000000  ...   
unique         NaN         NaN         NaN          NaN          NaN  ...   
top            NaN         NaN         NaN          NaN          NaN  ...   
freq           NaN         NaN         NaN          NaN          NaN  ...   
mean      2.458647    3.537594    5.390977     7.962406    11.003759  ...   
std      27.377862   34.083035   47.434934    66.289178    89.313757  ...   
min       0.000000    0.000000    0.000000     0.000000     0.000000  ...   
25%       0.000000    0.000000    0.000000     0.000000     0.000000  ...   
50%       0.000000    0.000000    0.000000     0.000000     0.000000  ...   
75%       0.000000    0.000000    0.000000     0.000000     0.000000  ...   
max     444.000000  549.000000  761.000000  1058.000000  1423.000000  ...   

              9/2/20        9/3/20        9/4/20        9/5/20        9/6/20  \
count   2.660000e+02  2.660000e+02  2.660000e+02  2.660000e+02  2.660000e+02   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    9.786244e+04  9.889044e+04  1.000676e+05  1.010584e+05  1.018942e+05   
std     5.143561e+05  5.197116e+05  5.260300e+05  5.316019e+05  5.361459e+05   
min     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%     4.517500e+02  4.517500e+02  4.540000e+02  4.610000e+02  4.625000e+02   
50%     2.928000e+03  2.975500e+03  3.022500e+03  3.072000e+03  3.113000e+03   
75%     3.096425e+04  3.105850e+04  3.115475e+04  3.127025e+04  3.134675e+04   
max     6.113510e+06  6.150016e+06  6.200518e+06  6.244970e+06  6.276365e+06   

              9/7/20        9/8/20        9/9/20       9/10/20       9/11/20  
count   2.660000e+02  2.660000e+02  2.660000e+02  2.660000e+02  2.660000e+02  
unique           NaN           NaN           NaN           NaN           NaN  
top              NaN           NaN           NaN           NaN           NaN  
freq             NaN           NaN           NaN           NaN           NaN  
mean    1.027735e+05  1.036494e+05  1.047469e+05  1.058701e+05  1.070730e+05  
std     5.399441e+05  5.442783e+05  5.497977e+05  5.556045e+05  5.620659e+05  
min     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%     4.645000e+02  4.652500e+02  4.652500e+02  4.660000e+02  4.667500e+02  
50%     3.153000e+03  3.186000e+03  3.228000e+03  3.273500e+03  3.320500e+03  
75%     3.139925e+04  3.151575e+04  3.181750e+04  3.234375e+04  3.345200e+04  
max     6.300622e+06  6.327009e+06  6.360212e+06  6.396100e+06  6.443743e+06  

[11 rows x 238 columns]

In [11]:
# Statistical checking for Global Deaths 
deaths.describe(include='all')

Province/State Country/Region         Lat        Long     1/22/20  \
count              81            266  266.000000  266.000000  266.000000   
unique             81            188         NaN         NaN         NaN   
top           Sichuan          China         NaN         NaN         NaN   
freq                1             33         NaN         NaN         NaN   
mean              NaN            NaN   21.076624   22.829441    0.063910   
std               NaN            NaN   24.904600   70.191330    1.042337   
min               NaN            NaN  -51.796300 -135.000000    0.000000   
25%               NaN            NaN    6.677575  -15.277675    0.000000   
50%               NaN            NaN   22.820850   20.921188    0.000000   
75%               NaN            NaN   41.143200   77.913685    0.000000   
max               NaN            NaN   71.706900  178.065000   17.000000   

           1/23/20     1/24/20     1/25/20     1/26/20     1/27/20  ...  \
count   266.000000  266.000000  266.000000  266.000000  266.000000  ...   
unique         NaN         NaN         NaN         NaN         NaN  ...   
top            NaN         NaN         NaN         NaN         NaN  ...   
freq           NaN         NaN         NaN         NaN         NaN  ...   
mean      0.067669    0.097744    0.157895    0.210526    0.308271  ...   
std       1.043908    1.473615    2.453621    3.189730    4.660845  ...   
min       0.000000    0.000000    0.000000    0.000000    0.000000  ...   
25%       0.000000    0.000000    0.000000    0.000000    0.000000  ...   
50%       0.000000    0.000000    0.000000    0.000000    0.000000  ...   
75%       0.000000    0.000000    0.000000    0.000000    0.000000  ...   
max      17.000000   24.000000   40.000000   52.000000   76.000000  ...   

               9/2/20         9/3/20         9/4/20         9/5/20  \
count      266.000000     266.000000     266.000000     266.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean      3244.466165    3265.913534    3288.116541    3306.680451   
std      15521.992322   15622.650068   15723.199898   15806.544108   
min          0.000000       0.000000       0.000000       0.000000   
25%          4.250000       5.000000       5.000000       5.000000   
50%         54.000000      54.000000      54.000000      55.000000   
75%        562.750000     563.500000     564.750000     565.250000   
max     185720.000000  186790.000000  187755.000000  188538.000000   

               9/6/20         9/7/20         9/8/20         9/9/20  \
count      266.000000     266.000000     266.000000     266.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean      3320.823308    3355.812030    3373.620301    3397.315789   
std      15860.914755   15913.901869   15981.316278   16099.587077   
min          0.000000       0.000000       0.000000       0.000000   
25%          5.000000       5.000000       5.000000       5.000000   
50%         55.500000      57.000000      57.000000      57.000000   
75%        566.000000     567.500000     570.000000     570.750000   
max     188941.000000  189208.000000  189653.000000  190859.000000   

              9/10/20        9/11/20  
count      266.000000     266.000000  
unique            NaN            NaN  
top               NaN            NaN  
freq              NaN            NaN  
mean      3419.093985    3441.187970  
std      16202.430513   16314.668589  
min          0.000000       0.000000  
25%          5.000000       5.000000  
50%         57.000000      57.000000  
75%        576.750000     582.750000  
max     191766.000000  192979.000000  

[11 rows x 238 columns]

In [12]:
# Checking Missing Value
print("GLOBAL CONFIRMED CASES")
cases.info()
print(cases.isnull().sum().sort_values(ascending=False))
print("\n GLOBAL DEATH CASES")
deaths.info()
print(deaths.isnull().sum().sort_values(ascending=False))

GLOBAL CONFIRMED CASES
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 266 entries, 0 to 265
Columns: 238 entries, Province/State to 9/11/20
dtypes: float64(2), int64(234), object(2)
memory usage: 494.7+ KB
Province/State    185
4/1/20              0
4/13/20             0
4/12/20             0
4/11/20             0
                 ... 
6/20/20             0
6/19/20             0
6/18/20             0
6/17/20             0
5/15/20             0
Length: 238, dtype: int64

 GLOBAL DEATH CASES
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 266 entries, 0 to 265
Columns: 238 entries, Province/State to 9/11/20
dtypes: float64(2), int64(234), object(2)
memory usage: 494.7+ KB
Province/State    185
4/1/20              0
4/13/20             0
4/12/20             0
4/11/20             0
                 ... 
6/20/20             0
6/19/20             0
6/18/20             0
6/17/20             0
5/15/20             0
Length: 238, dtype: int64


**Summary**

Missing values are in _Province/State_ columns for both Global Confirmed Cases & Deaths. Each file has 185 null values

#### 2. Data Cleaning & Consolidation

#### Quick Note
In here, I will use *pandas.melt* to unpivot a DataFrame from wide to long format where one or more cols are idenfier variables (*id_vars*), while all other columns are "unvipoted" to the row axis

In [13]:
# Example
df = pd.DataFrame({"A":{0:'a',1:'b',2:'c'},
                   "B":{0: 1, 1: 3, 2: 5},
                   "C":{0: 2, 1: 4, 2: 6}})
df

A  B  C
0  a  1  2
1  b  3  4
2  c  5  6

In [14]:
pd.melt(df, id_vars=['A'], value_vars=['B'])

A variable  value
0  a        B      1
1  b        B      3
2  c        B      5

In [15]:
pd.melt(df, id_vars=['A'], value_vars=['B','C'])

A variable  value
0  a        B      1
1  b        B      3
2  c        B      5
3  a        C      2
4  b        C      4
5  c        C      6

#### Let's get started

- Merge 2 datasets: confirmed cases & deaths
- Clean the values in 'Province/State'
- Renew the country name (package 'pycountry_convert')
- Fixing incorrect data in certain date in 'Hubei', China


**A. DATA CONSOLIDATION**

In [16]:
# Combine Confirmed & Death Cases
date = cases.columns[4:]
# unpivot Global Confirmed Cases dataset
cases_df = cases.melt(id_vars=['Province/State','Country/Region','Lat','Long'],
                      value_vars = date,
                      var_name = 'Date',
                      value_name = 'Confirmed Cases')

# unpivot Gloal Deaths Cases dataset
deaths_df = deaths.melt(id_vars=['Province/State', 'Country/Region','Lat','Long'],
                        value_vars = date,
                        var_name = 'Date',
                        value_name = 'Deaths')

# merged two new data frame
df = pd.concat([cases_df, deaths_df['Deaths']],
              axis = 1,
              sort = False)

# Check the data
print(df.columns)      # print out columns name
df.tail()

Index(['Province/State', 'Country/Region', 'Lat', 'Long', 'Date',
       'Confirmed Cases', 'Deaths'],
      dtype='object')


Province/State      Country/Region        Lat       Long     Date  \
62239            NaN  West Bank and Gaza  31.952200  35.233200  9/11/20   
62240            NaN      Western Sahara  24.215500 -12.885800  9/11/20   
62241            NaN               Yemen  15.552727  48.516388  9/11/20   
62242            NaN              Zambia -13.133897  27.849332  9/11/20   
62243            NaN            Zimbabwe -19.015438  29.154857  9/11/20   

       Confirmed Cases  Deaths  
62239            29256     204  
62240               10       1  
62241             2007     582  
62242            13323     306  
62243             7479     224

In [17]:
df.to_csv('merged_uncleaned_data.csv')

**B. DATA CLEANING**
- Columns Province/State

In [18]:
df['Province/State'].unique()

array([nan, 'Australian Capital Territory', 'New South Wales',
       'Northern Territory', 'Queensland', 'South Australia', 'Tasmania',
       'Victoria', 'Western Australia', 'Alberta', 'British Columbia',
       'Diamond Princess', 'Grand Princess', 'Manitoba', 'New Brunswick',
       'Newfoundland and Labrador', 'Northwest Territories',
       'Nova Scotia', 'Ontario', 'Prince Edward Island', 'Quebec',
       'Saskatchewan', 'Yukon', 'Anhui', 'Beijing', 'Chongqing', 'Fujian',
       'Gansu', 'Guangdong', 'Guangxi', 'Guizhou', 'Hainan', 'Hebei',
       'Heilongjiang', 'Henan', 'Hong Kong', 'Hubei', 'Hunan',
       'Inner Mongolia', 'Jiangsu', 'Jiangxi', 'Jilin', 'Liaoning',
       'Macau', 'Ningxia', 'Qinghai', 'Shaanxi', 'Shandong', 'Shanghai',
       'Shanxi', 'Sichuan', 'Tianjin', 'Tibet', 'Xinjiang', 'Yunnan',
       'Zhejiang', 'Faroe Islands', 'Greenland', 'French Guiana',
       'French Polynesia', 'Guadeloupe', 'Martinique', 'Mayotte',
       'New Caledonia', 'Reunion', 'Sai

In [19]:
len(df['Province/State'].unique())

82

In [20]:
# Correct the values in 'Province/State'
# Remove the ',' in Bonaire, Sint Eustatius and Saba to avoid redundancy
df = df[df['Province/State'].str.contains(',')!=True]

In [21]:
len(df['Province/State'].unique())

81

In [22]:
len(df['Country/Region'].unique())

188

In [23]:
df['Country/Region'].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Brunei', 'Bulgaria', 'Burkina Faso', 'Burma', 'Burundi',
       'Cabo Verde', 'Cambodia', 'Cameroon', 'Canada',
       'Central African Republic', 'Chad', 'Chile', 'China', 'Colombia',
       'Comoros', 'Congo (Brazzaville)', 'Congo (Kinshasa)', 'Costa Rica',
       "Cote d'Ivoire", 'Croatia', 'Cuba', 'Cyprus', 'Czechia', 'Denmark',
       'Diamond Princess', 'Djibouti', 'Dominica', 'Dominican Republic',
       'Ecuador', 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Eritrea',
       'Estonia', 'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France',
       'Gabon', 'Gambia', 'Georgia', 'Germany', 'Ghana', 'Greece',
       'Grenada', 'Guatemala', 'Guin

**C. Renew Country Name**

In [24]:
# Changing the country name for using the package 'pycountry_convert'
#!pip install pycountry-convert
# Dictionary of new country name
country_dict = {'Taiwan*': 'Taiwan',
          'US': 'USA',
          'Korea, South':'South Korea',
          'Reunion':'Réunion',
          'Congo (Kinshasa)': 'Democratic Republic of the Congo',
          'Congo (Brazzaville)': 'Republic of the Congo',
          "Cote d'Ivoire": "Côte d'Ivoire",
           'Bahamas, The': 'Bahamas',
           'Gambia, The': 'Gambia'
    }

for o,n in country_dict.items():
    df['Country/Region']=df['Country/Region'].replace(o,n)

In [25]:
print(df['Country/Region'].unique())
print(len(df['Country/Region'].unique()))

['Afghanistan' 'Albania' 'Algeria' 'Andorra' 'Angola'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Australia' 'Austria'
 'Azerbaijan' 'Bahamas' 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus'
 'Belgium' 'Belize' 'Benin' 'Bhutan' 'Bolivia' 'Bosnia and Herzegovina'
 'Botswana' 'Brazil' 'Brunei' 'Bulgaria' 'Burkina Faso' 'Burma' 'Burundi'
 'Cabo Verde' 'Cambodia' 'Cameroon' 'Canada' 'Central African Republic'
 'Chad' 'Chile' 'China' 'Colombia' 'Comoros' 'Republic of the Congo'
 'Democratic Republic of the Congo' 'Costa Rica' "Côte d'Ivoire" 'Croatia'
 'Cuba' 'Cyprus' 'Czechia' 'Denmark' 'Diamond Princess' 'Djibouti'
 'Dominica' 'Dominican Republic' 'Ecuador' 'Egypt' 'El Salvador'
 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Eswatini' 'Ethiopia' 'Fiji'
 'Finland' 'France' 'Gabon' 'Gambia' 'Georgia' 'Germany' 'Ghana' 'Greece'
 'Grenada' 'Guatemala' 'Guinea' 'Guinea-Bissau' 'Guyana' 'Haiti'
 'Holy See' 'Honduras' 'Hungary' 'Iceland' 'India' 'Indonesia' 'Iran'
 'Iraq' 'Ireland' 'Israel' 'Italy' 'Ja

In [26]:
# Checking the result
df[df['Country/Region']=='Taiwan']

Province/State Country/Region   Lat   Long     Date  Confirmed Cases  \
233              NaN         Taiwan  23.7  121.0  1/22/20                1   
499              NaN         Taiwan  23.7  121.0  1/23/20                1   
765              NaN         Taiwan  23.7  121.0  1/24/20                3   
1031             NaN         Taiwan  23.7  121.0  1/25/20                3   
1297             NaN         Taiwan  23.7  121.0  1/26/20                4   
...              ...            ...   ...    ...      ...              ...   
61147            NaN         Taiwan  23.7  121.0   9/7/20              494   
61413            NaN         Taiwan  23.7  121.0   9/8/20              495   
61679            NaN         Taiwan  23.7  121.0   9/9/20              495   
61945            NaN         Taiwan  23.7  121.0  9/10/20              496   
62211            NaN         Taiwan  23.7  121.0  9/11/20              498   

       Deaths  
233         0  
499         0  
765         0  
1031        0  
1297        0  
...       ...  
61147       7  
61413       7  
61679       7  
61945       7  
62211       7  

[234 rows x 7 columns]

**D. Fixing Incorrect Hubei Data On 2/12/2020**

In [27]:
df[(df['Province/State']=='Hubei')&(df['Date']=='2/12/20')]

Province/State Country/Region      Lat      Long     Date  \
5655          Hubei          China  30.9756  112.2707  2/12/20   

      Confirmed Cases  Deaths  
5655            33366    1068

In [28]:
# Create a function to correct the value
def correcting_value(region, date, value_name, correct_value):
    for key, val in correct_value.items():
        df.loc[(df[region]==key)&(df['Date']==date), value_name]=val

In [29]:
# Correct Data
hubei21220 = {'Hubei':34874}

# Correct the data
correcting_value('Province/State','2/12/20','Confirmed Cases', hubei21220)

In [30]:
# Checking the result
df[(df['Province/State']=='Hubei')&(df['Date']=='2/12/20')]

Province/State Country/Region      Lat      Long     Date  \
5655          Hubei          China  30.9756  112.2707  2/12/20   

      Confirmed Cases  Deaths  
5655            34874    1068

**E. Convert Date From String to Datetime Type**

In [31]:
# Convert date from string to datetime type using pd.to_datetime
# Use strftime to sort the date order again (month, date, year) 
df['Date']=pd.to_datetime(df['Date']).dt.strftime('%m/%d/%Y')
df.tail()

Province/State      Country/Region        Lat       Long        Date  \
62239            NaN  West Bank and Gaza  31.952200  35.233200  09/11/2020   
62240            NaN      Western Sahara  24.215500 -12.885800  09/11/2020   
62241            NaN               Yemen  15.552727  48.516388  09/11/2020   
62242            NaN              Zambia -13.133897  27.849332  09/11/2020   
62243            NaN            Zimbabwe -19.015438  29.154857  09/11/2020   

       Confirmed Cases  Deaths  
62239            29256     204  
62240               10       1  
62241             2007     582  
62242            13323     306  
62243             7479     224

**F. Save The Data For Further Analysis**

In [32]:
df.to_csv('covid19_clean_data.csv')

### Part IV: Data Analysis & Visualization


#### 1. Global Level: Cases & Deaths

In [33]:
# Extract the final date of the data frame
final_date = df['Date'].iloc[-1]
print(final_date)

most_recent_df = df[df['Date']==final_date]
print('Globally COVID-19 Information on {}:\n'.format(final_date))
print('Confirmed Cases: {:,}'.format(most_recent_df['Confirmed Cases'].sum()))
print('Death Cases: {:,}'.format(most_recent_df['Deaths'].sum()))

09/11/2020
Globally COVID-19 Information on 09/11/2020:

Confirmed Cases: 28,481,388
Death Cases: 915,356


#### 2. Country Level: Cases & Deaths

In [34]:
# Color pallete for cmap is diverging orange red
color_confirmedcases = 'OrRd'
color_death = 'OrRd'

# Extract the data by country on the most recent date
country_df = most_recent_df.groupby('Country/Region')['Confirmed Cases','Deaths'].sum().reset_index()

# Create a highlighting color on Country Cases
country_df.sort_values('Confirmed Cases', ascending=False)\
    .style.background_gradient(cmap=color_confirmedcases, subset=['Confirmed Cases'])\
    .background_gradient(cmap=color_death, subset=['Deaths'])                           

**3. Top 10 Country COVID-19 Chart**

In [35]:
# Prep data for the graph: Extracting top 10 countries
top10_country = country_df.sort_values(['Confirmed Cases'], ascending=False).head(10)

# Using Altair (alt) for this single value graph
# Parameter list
bar_color = '#076C8C' # HTML color coding
sort_value='Confirmed Cases'
y_axis = 'Country/Region' # Province/State
top_num=10
threshold = 1000

# Create the bar chart
chart = alt.Chart(top10_country)\
            .mark_bar(color=bar_color,     
                     size=20,               
                     opacity=0.7)\
            .encode(
            x= alt.X('{}:Q'.format(sort_value), title = 'Confirmed Cases'),
            y= alt.Y('{}:O'.format(y_axis), title='Country/Region', sort='-x'))\
            .properties(
            title = {
            "text":['Top {}: {}'.format(y_axis,sort_value)],
            "subtitle":['Data is updated on {}'.format(final_date)],
            "fontSize": 17,
            "fontWeight":'bold',
            "font":'Courier New'
            })

# Add labels to the chart
text = chart.mark_text(align = 'left', baseline='middle', dx=3)\
           .encode(text = sort_value)

# Show the chart + label
(chart+text).properties(height=375, width=700)

alt.LayerChart(...)

Using **Plotly** to map the top 10 countries with the **HIGHEST CONFIRMED CASES**

In [36]:
# Using Plotly
fig = px.bar(top10_country,
             x = 'Country/Region',
             y = 'Confirmed Cases',
             #text = 'Confirmed Cases',  
             hover_name = 'Country/Region',
             hover_data = ['Confirmed Cases'])

# Adding title to the plot
fig.update_layout(title = 'Top 10 Country/Region: Confirmed Cases',
                 xaxis_tickfont_size=12,
                 yaxis_tickfont_size=12,
                 bargap=0.25)
fig.show()

Top 10 Countries with the **HIGHEST DEATH**

In [37]:
# Using Plotly
top10_deaths = country_df.sort_values(['Deaths'], ascending=False).head(10)
fig = px.bar(top10_deaths,
             x = 'Country/Region',
             y = 'Deaths',
             #text = 'Confirmed Cases',  
             hover_name = 'Country/Region',
             hover_data = ['Deaths'])

# Adding title to the plot
fig.update_layout(title = 'Top 10 Country/Region: Deaths',
                 xaxis_tickfont_size=12,
                 yaxis_tickfont_size=12,
                 bargap=0.25)
fig.show()

In [38]:
import plotly.graph_objects as go
# Using Plotly
fig = go.Figure()
fig.add_trace(go.Bar(x = top10_country['Country/Region'],
                     y = top10_country['Confirmed Cases'],
                     name = 'Confirmed Cases',
                     marker_color = 'indianred'))

fig.add_trace(go.Bar(x = top10_country['Country/Region'],
                    y = top10_country['Deaths'],
                    name = 'Deaths',
                    marker_color='lightsalmon'))

fig.update_layout(barmode='group', xaxis_tickangle = -45)
fig.show()

In [39]:
top10_country

Country/Region  Confirmed Cases  Deaths
174            USA          6443743  192979
78           India          4659984   77472
23          Brazil          4282164  130396
139         Russia          1048257   18309
132           Peru           710067   30344
37        Colombia           702088   22518
111         Mexico           658299   70183
155   South Africa           646398   15378
158          Spain           566326   29747
6        Argentina           535705   11148

**4. Geographical map of cases by province**

In [40]:
# Using most recent data
most_recent_df.head()

Province/State Country/Region       Lat       Long        Date  \
61978            NaN    Afghanistan  33.93911  67.709953  09/11/2020   
61979            NaN        Albania  41.15330  20.168300  09/11/2020   
61980            NaN        Algeria  28.03390   1.659600  09/11/2020   
61981            NaN        Andorra  42.50630   1.521800  09/11/2020   
61982            NaN         Angola -11.20270  17.873900  09/11/2020   

       Confirmed Cases  Deaths  
61978            38606    1420  
61979            11021     327  
61980            47752    1599  
61981             1344      53  
61982             3279     131

In [41]:
# Using folium
# Data: Using most_recent_data dataframe

## parameter for map
mapstyle = 'CartoDB positron'
line_color = '#da635eff'
fill_color = '#da635eff'
fill_opacity = 0.6
# other styles: 'OpenStreetMap', "Stamen Terrain”, “Stamen Toner”, “Stamen Watercolor”

## create map
world_map = folium.Map(location = [10,0], zoom_start = 2, max_zoom = 8, min_zoom = 2, tiles = mapstyle)
## define detail of the map
for lat, long, case, name in zip(most_recent_df['Lat'], most_recent_df['Long'], most_recent_df['Confirmed Cases'],\
                                most_recent_df['Country/Region']):
    folium.CircleMarker([lat, long], radius = (int((np.log(case+1.00001)))+0.2),
                       popup = ("<h5 style='text-align:center;font-weight: bold'>" + str(name).capitalize()+ "</h5>" + '<br>'
                                '<strong>Confirmed Cases</strong>: ' + str(case) + '<br>'),\
                       color = line_color, weight= 1.5, \
                        fill_color = fill_color, fill_opacity = fill_opacity).add_to(world_map)
# opacity = fill_opacity #opacity of the line 
 
## Save map
world_map.save("./world_map.html")
world_map


Next, global COVID-19 confirmed cases are plotted on **Choropleth Maps**.

A **Choropleth Map** is a map composed of colored polygons, which is used to represent spacial variations of a quantity.

Base map configuration `px.choropleth (data, locations, color, color_continuous_scale, range_color, scope, labels)`

[For further information on Choropleth, click here](https://plotly.com/python/choropleth-maps/)

In [42]:
#### ****** MAP - STYLE 2 ******
# Using choropleth
# Data: using country_df dataframe (log10 scale)

# list of parameters
color = '#da635eff' # Red
map_value = 'Confirmed Cases'  # 'Deaths'/'Recovered'

# Create a function to plot the map
def plotting_global_heatmap(country_df, map_value, color):
    temp = country_df[['Country/Region',map_value]]
    plot = px.choropleth(temp, locations='Country/Region',   # map configuration
                        color = np.log10(temp[map_value] + 1), # + 1 to avoid divided by 0, log0
                        hover_name = 'Country/Region', # set hover name
                        hover_data = [map_value], 
                        color_continuous_scale = px.colors.sequential.Plasma, locationmode = "country names")
    plot.update_layout(title_text = "{} Heat Map (Log Scale)".format(map_value), title_x=0.5)
    plot.update_coloraxes(colorbar_title="{}(Log Scale)".format(map_value), colorscale = "Reds")
    return plot

# Plotting global heat map
plot = plotting_global_heatmap(country_df,map_value,color)
#plot.to_image("Global Heal Map.png")
plot.show()

In [43]:
#### ****** MAP - STYLE 2 ******
# Using choropleth
# Data: using country_df dataframe (log10 scale)

# list of parameters
color = 'Greys' # Red
map_value = 'Deaths'  # 'Deaths'/'Recovered'

# Create a function to plot the map
def plotting_global_heatmap(country_df, map_value, color):
    temp = country_df[['Country/Region',map_value]]
    plot = px.choropleth(temp, locations='Country/Region',   # map configuration
                        color = np.log10(temp[map_value] + 1), # + 1 to avoid divided by 0, log0
                        hover_name = 'Country/Region', # set hover name
                        hover_data = [map_value], 
                        color_continuous_scale = px.colors.sequential.Greys, locationmode = "country names")
    plot.update_layout(title_text = "{} Heat Map (Log Scale)".format(map_value), title_x=0.5)
    plot.update_coloraxes(colorbar_title="{}(Log Scale)".format(map_value), colorscale = "Greys")
    return plot

# Plotting global heat map
plot = plotting_global_heatmap(country_df,map_value,color)
#plot.to_image("Global Heal Map.png")
plot.show()

**5. Worldwide Detail Report Since The Early Stage of The Pandemic - Up to Now**

In [44]:
# Using df data frames
# Note: Some of the Country/Region has details about Province/State so if analyzing country data we must use groupby('Country/Region') to sum the cases all the Province/State of the same country
# For example, Canada is :
df[df['Country/Region']=='Canada']

Province/State Country/Region      Lat      Long        Date  \
39                  Alberta         Canada  53.9333 -116.5765  01/22/2020   
40         British Columbia         Canada  53.7267 -127.6476  01/22/2020   
41         Diamond Princess         Canada   0.0000    0.0000  01/22/2020   
42           Grand Princess         Canada   0.0000    0.0000  01/22/2020   
43                 Manitoba         Canada  53.7609  -98.8139  01/22/2020   
...                     ...            ...      ...       ...         ...   
62026               Ontario         Canada  51.2538  -85.3232  09/11/2020   
62027  Prince Edward Island         Canada  46.5107  -63.4168  09/11/2020   
62028                Quebec         Canada  52.9399  -73.5491  09/11/2020   
62029          Saskatchewan         Canada  52.9399 -106.4509  09/11/2020   
62030                 Yukon         Canada  64.2823 -135.0000  09/11/2020   

       Confirmed Cases  Deaths  
39                   0       0  
40                   0       0  
41                   0       0  
42                   0       0  
43                   0       0  
...                ...     ...  
62026            46118    2863  
62027               55       0  
62028            64463    5774  
62029             1688      24  
62030               15       0  

[3276 rows x 7 columns]

**Static Graph**

In [45]:
# Sum up all the cases in the world by date
world_cases_all_time = df.groupby('Date')['Confirmed Cases','Deaths'].sum().sort_values('Date').reset_index()

# Adding new confirmed cases which indicates the difference between day of confirmed cases
# For example - 99 is the number of cases increased from 555 in 2020-1-22 to 654 in 2020-1-23

world_cases_all_time['New Cases'] = world_cases_all_time['Confirmed Cases'] - world_cases_all_time['Confirmed Cases'].shift(1)

# Check the adding
world_cases_all_time.head()
#world_cases_all_time.tail()

Date  Confirmed Cases  Deaths  New Cases
0  01/22/2020              555      17        NaN
1  01/23/2020              654      18       99.0
2  01/24/2020              941      26      287.0
3  01/25/2020             1434      42      493.0
4  01/26/2020             2118      56      684.0

**Line Graph**

In [46]:
# For good x-axis display, show only month & year
world_cases_all_time['Date'] = pd.to_datetime(world_cases_all_time['Date'])
#world_cases_all_time['Date'] = world_cases_all_time['Date'].dt.strftime('%m/%d')

In [47]:
world_cases_all_time

Date  Confirmed Cases  Deaths  New Cases
0   2020-01-22              555      17        NaN
1   2020-01-23              654      18       99.0
2   2020-01-24              941      26      287.0
3   2020-01-25             1434      42      493.0
4   2020-01-26             2118      56      684.0
..         ...              ...     ...        ...
229 2020-09-07         27337742  892646   233915.0
230 2020-09-08         27570721  897383   232979.0
231 2020-09-09         27862659  903686   291938.0
232 2020-09-10         28161409  909479   298750.0
233 2020-09-11         28481388  915356   319979.0

[234 rows x 4 columns]

In [48]:
# Pivot the table
world_cases_melt = world_cases_all_time.melt(id_vars=['Date'],
                         value_vars = ['Confirmed Cases', 'Deaths', 'New Cases'])
world_cases_melt.head()

# plotting the cases trend line chart
plot = px.line(world_cases_melt, x='Date',y='value', color='variable')
plot.update_layout(title={'text': 'Global Confirmed Cases & Deaths Over Time', 'x':0.8},
                   xaxis_title = None,
                   yaxis_title = 'Cases',
                   legend={'title': None})
plot.show()

In [49]:
world_cases_melt

Date         variable     value
0   2020-01-22  Confirmed Cases     555.0
1   2020-01-23  Confirmed Cases     654.0
2   2020-01-24  Confirmed Cases     941.0
3   2020-01-25  Confirmed Cases    1434.0
4   2020-01-26  Confirmed Cases    2118.0
..         ...              ...       ...
697 2020-09-07        New Cases  233915.0
698 2020-09-08        New Cases  232979.0
699 2020-09-09        New Cases  291938.0
700 2020-09-10        New Cases  298750.0
701 2020-09-11        New Cases  319979.0

[702 rows x 3 columns]

**Bar Graph**

In [50]:
# Using the data frame world_cases_all_time
plot = go.Figure(data=[
    go.Bar(name='Confirmed Cases', 
           x=world_cases_all_time['Date'],
           y=world_cases_all_time['Confirmed Cases']),
    go.Bar(name='Deaths',
          x=world_cases_all_time['Date'],
          y=world_cases_all_time['Deaths'])
])

plot.update_layout(title={'text': 'Global Confirmed Cases & Deaths Over Time', 'x':0.8},
                  xaxis_title=None,
                  yaxis_title='Cases',
                  legend={'title':None})

plot.show()

**Animated Graph**

In [51]:
# Sum up all the cases for each country by date
countries_cases_all_time = df.groupby(['Date','Country/Region'])['Confirmed Cases', 'Deaths'].sum().sort_values('Date').reset_index()

# Check the grouping
countries_cases_all_time.tail()

Date Country/Region  Confirmed Cases  Deaths
43987  09/11/2020        Georgia             1917      19
43988  09/11/2020        Germany           259735    9348
43989  09/11/2020          Ghana            45388     285
43990  09/11/2020       Eswatini             5025      98
43991  09/11/2020       Zimbabwe             7479     224

In [52]:
world_cases_all_time

Date  Confirmed Cases  Deaths  New Cases
0   2020-01-22              555      17        NaN
1   2020-01-23              654      18       99.0
2   2020-01-24              941      26      287.0
3   2020-01-25             1434      42      493.0
4   2020-01-26             2118      56      684.0
..         ...              ...     ...        ...
229 2020-09-07         27337742  892646   233915.0
230 2020-09-08         27570721  897383   232979.0
231 2020-09-09         27862659  903686   291938.0
232 2020-09-10         28161409  909479   298750.0
233 2020-09-11         28481388  915356   319979.0

[234 rows x 4 columns]

#### F. Country Comparison

In [53]:
# Sum up all cases for each country by date
countries_cases_all_time = df.groupby(['Date','Country/Region'])['Confirmed Cases','Deaths'].sum().sort_values('Date').reset_index()

# Checking the grouping
countries_cases_all_time.tail()

Date Country/Region  Confirmed Cases  Deaths
43987  09/11/2020        Georgia             1917      19
43988  09/11/2020        Germany           259735    9348
43989  09/11/2020          Ghana            45388     285
43990  09/11/2020       Eswatini             5025      98
43991  09/11/2020       Zimbabwe             7479     224

In [54]:
# Top 10 country
sort_value = 'Confirmed Cases'

# Figure out which top 10 countries with the highest number of confirmed cases
top10_country = country_df.sort_values(sort_value,ascending=False).head(10)['Country/Region'].unique()
print(top10_country)

# Extract top 10 countries cases from countries cases all time
top10_country_cases = countries_cases_all_time[countries_cases_all_time['Country/Region'].isin(top10_country)]
top10_country_cases['Date']=pd.to_datetime(top10_country_cases['Date'])
top10_country_cases

['USA' 'India' 'Brazil' 'Russia' 'Peru' 'Colombia' 'Mexico' 'South Africa'
 'Spain' 'Argentina']


Date Country/Region  Confirmed Cases  Deaths
13    2020-01-22           Peru                0       0
33    2020-01-22         Russia                0       0
39    2020-01-22         Mexico                0       0
54    2020-01-22            USA                1       0
83    2020-01-22   South Africa                0       0
...          ...            ...              ...     ...
43891 2020-09-11          Spain           566326   29747
43899 2020-09-11         Brazil          4282164  130396
43913 2020-09-11       Colombia           702088   22518
43929 2020-09-11      Argentina           535705   11148
43953 2020-09-11          India          4659984   77472

[2340 rows x 4 columns]

In [63]:
# Plot top 10 country
fig = px.line(top10_country_cases,
              x='Date',
              y='Confirmed Cases',
              color='Country/Region',
              title = f'Confirmed Cases for Top 10 Country')
fig.update_layout(legend={'title': None})
fig.show()

In [58]:
fig = px.line(top10_country_cases,
              x='Date',
              y='Confirmed Cases',
              color='Country/Region',
              title=f'Confirmed Cases for Top 10 Country (Log)')
fig.update_layout(legend={'title':None},
                 xaxis_title='Date',
                 yaxis_title='Log(cases)')
fig.update_layout(yaxis_type="log")
fig.show()